File: Hash/베스트앨범.py
from functools import reduce

def solution(genres, plays):
    answer = []
    
    by_genre = {}
    
    for i, (genre, play) in enumerate(zip(genres, plays)):
        if genre in by_genre: by_genre[genre].append((i, play))
        else: by_genre[genre] = [(i, play)]
    
    for genre in by_genre:
        by_genre[genre].sort(key = lambda x: (-x[1], x[0]))
    
    # use reduce
    plays_per_genre = {key: reduce(lambda x, y: x + y[1], by_genre[key], 0) for key in by_genre}
    plays_per_genre = [key for key, value in sorted(plays_per_genre.items(), key = lambda x: -x[1])]
    
    # or sort by sum
    # plays_per_genre = sorted(list(by_genre.keys()), key = lambda x: sum(map(lambda y: y[1], by_genre[x])), reverse = True)
    
    for genre in plays_per_genre:
        answer += [i for i, play in by_genre[genre]][:2]
    
    return answer

File: Hash/베스트앨범.js
function solution(genres, plays) {
    var answer = [];
    const map = new Map();
    const list = new Map();
    
    genres.forEach((_, i) => {
        if (!map.get(genres[i])) {
            map.set(genres[i], plays[i]);
            list.set(genres[i], [{num:plays[i], index:i}]);
        } else {
            map.set(genres[i], map.get(genres[i]) + plays[i]);
            list.set(genres[i], list.get(genres[i])
                     .concat([{num:plays[i], index:i}]));
        }
    });
    
    let arr = [...map.entries()];
    arr = arr.sort((a,b) => b[1] - a[1])
    arr.forEach((item) => {
        let top2 = list.get(item[0]).sort((a,b) => b.num - a.num).slice(0, 2);
        top2.forEach((t) => {
            answer = answer.concat(t.index);
        });
    });
    
    return answer;
}
